internal/apps/dop/providers/queue/calculate.go

```diff
@@ -0,0 +1,92 @@
+// Copyright (c) 2021 Terminus, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package queue
+
+import (
+	"fmt"
+
+	"github.com/alecthomas/units"
+
+	"github.com/erda-project/erda/apistructs"
+)
+
+type ProjectQueueResource struct {
+	Concurrency int64
+	MaxCPU      float64
+	MaxMemoryMB float64
+}
+
+func maxInt64(a, b int64) int64 {
+	if a >= b {
+		return a
+	}
+	return b
+}
+
+func maxFloat64(a, b float64) float64 {
+	if a >= b {
+		return a
+	}
+	return b
+}
+
+// calculateConcurrency calculates the queue concurrency for the given project.
+// each application is allocated two concurrency
+func (p *provider) calculateConcurrency(project *apistructs.ProjectDTO) (int64, error) {
+	apps, err := p.bdl.CountAppByProID(project.ID)
+	if err != nil {
+		return 0, err
+	}
+	concurrency := apps * 2
+	return maxInt64(concurrency, p.Cfg.DefaultQueueConcurrency), nil
+}
+
+func (p *provider) calculateMaxCpu(resourceConfig *apistructs.ResourceConfigInfo) (float64, error) {
+	return maxFloat64(p.Cfg.DefaultQueueCPU, resourceConfig.CPUQuota), nil
```

Comment:
I think you should compare with 0 if no resource available.

Reply:
compared